### DISCUSSION TITLE
Can i know the content is edited?

### DISCUSSION THREAD

**Discussion by chenxiaoleizi** [ueberdosis/tiptap#5430] [General Questions] [answered]
When user leave the editor page, if the user edit the content, and not saved to the server side, i want to give the user a message tell them leave will abord the edit part? so how can i know the content is edited?

**Comment by nperez0111** [comment]
Depends on if you have initial content, if you have initial content then you could use `editor.isEmpty` to know that it has content that should probably be saved back.

If you had initial content you could compare the initial content to the modified version and see if they are different.

Or the simplest might even be to set a flag if there ever is an `update` event on the editor (which indicates a modification to the document).

So there are many ways to tell this

**Reply by chenxiaoleizi** [reply]
Thanks for you replay. But there is a special situation, user edit the content first, then undo them?

**Reply by nperez0111** [reply]
That was the second statement
> If you had initial content you could compare the initial content to the modified version and see if they are different.

**Reply by chenxiaoleizi** [reply]
The second statement not work form me, because even if i did not edit the content, the content in the editor will be added some p element.
I use `insertContentAt` method to init my content., i think i can use the third statement. But one invoke  `insertContentAt` sometimes trigger multiple `update` event. 
![code](https://github.com/user-attachments/assets/b30f4213-04ac-45f9-bef6-a09f2a128e87)

So i can not know which is triggered by `insertContentAt` , which is triggered by user edit. Is there a way to know which `update` event trigged by `insertContentAt` ?

**Reply by nperez0111** [reply]
Shouldn't your insertContentAt just be valid content (i.e. what the editor transforms it to) and then still use that for comparison.

**Reply by nperez0111** [reply]
The editor inserts a `p` tag because that is what the schema defines as valid content: https://tiptap.dev/docs/editor/core-concepts/schema

**Reply by chenxiaoleizi** [reply]
My raw content is transformed by AI, so it is not easy to get the valid content. But your replay inspire me that , i can get the init content after `insertContentAt` invoked (not use the raw content),  so when i need to compare, i can get the editor current content to compare with the content got after  `insertContentAt` invoked.

**Reply by nperez0111** [reply, marked as answer]
Yea that could do it.

You can also do the parsing before it goes into the editor (to see what the editor would create) and use that as the comparison point.

I've written this before:
```ts
import { createNodeFromContent, Editor } from '@tiptap/core'

export function tryParseToTiptapJSON(content: string, editor: Editor) {

  try {
    const nodeOrFragment = createNodeFromContent(content, editor.schema, {
      parseOptions: {
        preserveWhitespace: false,
      },
    })

    return nodeOrFragment
  } catch (e) {
    return null
  }
}
```

Call `.toJSON()` to get the JSON version if you don't want fragments or nodes

**Reply by chenxiaoleizi** [reply]
Thanks! I will try it.
